# -*- coding: utf-8 -*-
from setuptools import setup

packages = \
['geneci']

package_data = \
{'': ['*']}

install_requires = \
['docker>=6.1.3,<7.0.0',
 'iteround>=1.0.4,<2.0.0',
 'numpy>=1.26.4,<2.0.0',
 'pandas>=1.5.1,<2.0.0',
 'plotly>=5.11.0,<6.0.0',
 'scipy>=1.10.2,<2.0.0',
 'typer[all]>=0.6,<0.7']

entry_points = \
{'console_scripts': ['geneci = geneci.main:app']}

setup_kwargs = {
    'name': 'geneci',
    'version': '2.5.1',
    'description': 'PBEvoGen is a software package derived from MO-GENECI that incorporates a preference selection mechanism that allows the user to guide the exploration of the evolutionary algorithm based on prior knowledge of the search space.',
    'long_description': '![CI](https://github.com/AdrianSeguraOrtiz/GENECI/actions/workflows/ci.yml/badge.svg)\n![Release](https://github.com/AdrianSeguraOrtiz/GENECI/actions/workflows/release.yml/badge.svg)\n![Pypi](https://img.shields.io/pypi/v/GENECI/2.5.1)\n\n[PBEvoGen (Preference Based Evolutionary Gene network consensus inference)](https://github.com/AdrianSeguraOrtiz/PBEvoGen) is a software package derived from [MO-GENECI](https://github.com/AdrianSeguraOrtiz/MO-GENECI) that incorporates a preference selection mechanism that allows the user to guide the exploration of the evolutionary algorithm based on prior knowledge of the search space.\n\n![Alt text](https://github.com/AdrianSeguraOrtiz/GENECI/raw/v-2.5.1/docs/diagram.svg)\n\n\n# Prerequisites\n\n- Python => 3.9\n- Docker\n\n# Instalation\n\n```sh\npip install geneci==2.5.1\n```\n\n# Integrated techniques\n\nPBEvoGen integrates the 26 inference techniques already contemplated in [MO-GENECI](https://github.com/AdrianSeguraOrtiz/MO-GENECI): ARACNE, BC3NET, C3NET, CLR, GENIE3_RF, GRNBOOST2, GENIE3_ET, MRNET, MRNETB, PCIT, TIGRESS, KBOOST, MEOMI, JUMP3, NARROMI, CMI2NI, RSNET, PCACMI, LOCPCACMI, PLSNET, PIDC, PUC, GRNVBEM, LEAP, NONLINEARODES, INFERELATOR.\n\n# Example procedure\n\n1. **Obtain simulated expression data and their respective gold standards**. As in [MO-GENECI](https://github.com/AdrianSeguraOrtiz/MO-GENECI).\n\n2. **Inference and consensus** of networks for the selected expression data. Unlike in [MO-GENECI](https://github.com/AdrianSeguraOrtiz/MO-GENECI), PBEvoGen integrates the `--reference-point` parameter that allows the specification of a reference point to condition the selection of individuals.\n\n- **Form 1**: Procedure prefixed by the command run.\n\n```sh\ngeneci run --expression-data input_data/DREAM4/EXP/dream4_100_01_exp.csv \\\n           --technique ARACNE --technique BC3NET --technique C3NET --technique CLR \\\n           --technique GENIE3_RF --technique GRNBOOST2 --technique GENIE3_ET \\\n           --technique MRNET --technique MRNETB --technique PCIT --technique TIGRESS \\\n           --technique KBOOST --technique MEOMI --technique NARROMI --technique CMI2NI \\\n           --technique RSNET --technique PCACMI --technique LOCPCACMI --technique PLSNET \\\n           --technique PIDC --technique PUC --technique GRNVBEM --technique LEAP \\\n           --technique NONLINEARODES --technique INFERELATOR \\\n           --crossover-probability 0.9 --mutation-probability 0.05 --population-size 100 \\\n           --num-parents 3 --mutation-strength 0.1 \\\n           --num-evaluations 50000 --cut-off-criteria PercLinksWithBestConf --cut-off-value 0.4 \\\n           --function Quality --function DegreeDistribution --function Motifs \\\n           --algorithm NSGAII --plot-fitness-evolution --plot-pareto-front \\\n           --plot-parallel-coordinates --output-dir inferred_networks \\\n           --reference-point "0.65;35.0;-1800.0"\n```\n\n- **Form 2**: Division of the procedure into several commands\n\n```sh\n# 1. Inference using individual techniques\ngeneci infer-network --expression-data input_data/DREAM4/EXP/dream4_100_01_exp.csv \\\n                     --technique ARACNE --technique BC3NET --technique C3NET --technique CLR \\\n                     --technique GENIE3_RF --technique GRNBOOST2 --technique GENIE3_ET \\\n                     --technique MRNET --technique MRNETB --technique PCIT --technique TIGRESS \\\n                     --technique KBOOST --technique MEOMI --technique NARROMI --technique CMI2NI \\\n                     --technique RSNET --technique PCACMI --technique LOCPCACMI --technique PLSNET \\\n                     --technique PIDC --technique PUC --technique GRNVBEM --technique LEAP \\\n                     --technique NONLINEARODES --technique INFERELATOR \\\n                     --output-dir inferred_networks/geneci_consensus\n\n# 2. Optimize the assembly of the trust lists resulting from the above command\ngeneci optimize-ensemble --confidence-list inferred_networks/dream4_100_01_exp/lists/GRN_LOCPCACMI.csv \\\n                         --confidence-list inferred_networks/dream4_100_01_exp/lists/GRN_BC3NET.csv \\\n                         --confidence-list inferred_networks/dream4_100_01_exp/lists/GRN_PLSNET.csv \\\n                         --confidence-list inferred_networks/dream4_100_01_exp/lists/GRN_GRNVBEM.csv \\\n                         --confidence-list inferred_networks/dream4_100_01_exp/lists/GRN_CMI2NI.csv \\\n                         --confidence-list inferred_networks/dream4_100_01_exp/lists/GRN_CLR.csv \\\n                         --confidence-list inferred_networks/dream4_100_01_exp/lists/GRN_INFERELATOR.csv \\\n                         --confidence-list inferred_networks/dream4_100_01_exp/lists/GRN_GRNBOOST2.csv \\\n                         --confidence-list inferred_networks/dream4_100_01_exp/lists/GRN_PCACMI.csv \\\n                         --confidence-list inferred_networks/dream4_100_01_exp/lists/GRN_MRNET.csv \\\n                         --confidence-list inferred_networks/dream4_100_01_exp/lists/GRN_PCIT.csv \\\n                         --confidence-list inferred_networks/dream4_100_01_exp/lists/GRN_KBOOST.csv \\\n                         --confidence-list inferred_networks/dream4_100_01_exp/lists/GRN_MEOMI.csv \\\n                         --confidence-list inferred_networks/dream4_100_01_exp/lists/GRN_NONLINEARODES.csv \\\n                         --confidence-list inferred_networks/dream4_100_01_exp/lists/GRN_GENIE3_ET.csv \\\n                         --confidence-list inferred_networks/dream4_100_01_exp/lists/GRN_NARROMI.csv \\\n                         --confidence-list inferred_networks/dream4_100_01_exp/lists/GRN_GENIE3_RF.csv \\\n                         --confidence-list inferred_networks/dream4_100_01_exp/lists/GRN_RSNET.csv \\\n                         --confidence-list inferred_networks/dream4_100_01_exp/lists/GRN_PIDC.csv \\\n                         --confidence-list inferred_networks/dream4_100_01_exp/lists/GRN_ARACNE.csv \\\n                         --confidence-list inferred_networks/dream4_100_01_exp/lists/GRN_MRNETB.csv \\\n                         --confidence-list inferred_networks/dream4_100_01_exp/lists/GRN_TIGRESS.csv \\\n                         --confidence-list inferred_networks/dream4_100_01_exp/lists/GRN_LEAP.csv \\\n                         --confidence-list inferred_networks/dream4_100_01_exp/lists/GRN_PUC.csv \\\n                         --confidence-list inferred_networks/dream4_100_01_exp/lists/GRN_C3NET.csv \\\n                         --crossover-probability 0.9 --mutation-probability 0.05 --population-size 100 \\\n                         --num-parents 3 --mutation-strength 0.1 \\\n                         --num-evaluations 50000 --cut-off-criteria PercLinksWithBestConf --cut-off-value 0.4 \\\n                         --function Quality --function DegreeDistribution --function Motifs \\\n                         --algorithm NSGAII --plot-fitness-evolution --plot-pareto-front \\\n                         --plot-parallel-coordinates --output-dir inferred_networks/geneci_consensus \\\n                         --reference-point "0.65;35.0;-1800.0"\n```\n\n3. **Representation** of inferred networks. As in [MO-GENECI](https://github.com/AdrianSeguraOrtiz/MO-GENECI).\n4. **Evaluation** of the quality of the inferred gene network with respect to the gold standard. As in [MO-GENECI](https://github.com/AdrianSeguraOrtiz/MO-GENECI).\n5. **Binarization** of the inferred gene network. As in [MO-GENECI](https://github.com/AdrianSeguraOrtiz/MO-GENECI).\n\n# `geneci`\n\n**Usage**:\n\n```console\n$ geneci [OPTIONS] COMMAND [ARGS]...\n```\n\n**Options**:\n\n* `--install-completion`: Install completion for the current shell.\n* `--show-completion`: Show completion for the current shell, to copy it or customize the installation.\n* `--help`: Show this message and exit.\n\n**Commands**:\n\n* `apply-cut`: Converts a list of confidence values into a binary matrix that represents the final gene network.\n* `cluster-network`: Divide an initial gene network into several communities following the Infomap (recommended) or Louvain grouping algorithm.\n* `draw-network`: Draw gene regulatory networks from confidence lists.\n* `evaluate`: Evaluate the accuracy of the inferred network with respect to its gold standard.\n* `extract-data`: Extract public data generated by simulators such as SynTReN, Rogers and GeneNetWeaver, as well as data from known challenges like DREAM3, DREAM4, DREAM5 and IRMA.\n* `generate-data`: Simulate time series with gene expression levels using the SysGenSIM simulator. They can be generated from scratch or based on the interactions of a real gene network.\n* `infer-network`: Infer gene regulatory networks from expression data. Several techniques are available: ARACNE, BC3NET, C3NET, CLR, GENIE3_RF, GRNBOOST2, GENIE3_ET, MRNET, MRNETB, PCIT, TIGRESS, KBOOST, MEOMI, JUMP3, NARROMI, CMI2NI, RSNET, PCACMI, LOCPCACMI, PLSNET, PIDC, PUC, GRNVBEM, LEAP, NONLINEARODES and INFERELATOR.\n* `optimize-ensemble`: Analyzes several trust lists and builds a consensus network by applying an evolutionary algorithm.\n* `run`: Infer gene regulatory network from expression data by employing multiple unsupervised learning techniques and applying a genetic algorithm for consensus optimization.\n* `weighted-confidence`: Calculate the weighted sum of the confidence levels reported in several files based on a given distribution of weights.\n\nOnly modified command specifications are shown with respect to [MO-GENECI](https://github.com/AdrianSeguraOrtiz/MO-GENECI).\n\n## `geneci run`\n\nInfer gene regulatory network from expression data by employing multiple unsupervised learning techniques and applying a genetic algorithm for consensus optimization.\n\n**Usage**:\n\n```console\n$ geneci run [OPTIONS]\n```\n\n**Options**:\n\n* `--expression-data PATH`: Path to the CSV file with the expression data. Genes are distributed in rows and experimental conditions (time series) in columns.  [required]\n* `--time-series PATH`: Path to the CSV file with the time series from which the individual gene networks have been inferred. This parameter is only necessary in case of specifying the fitness function Loyalty.\n* `--technique [ARACNE|BC3NET|C3NET|CLR|GENIE3_RF|GRNBOOST2|GENIE3_ET|MRNET|MRNETB|PCIT|TIGRESS|KBOOST|MEOMI|JUMP3|NARROMI|CMI2NI|RSNET|PCACMI|LOCPCACMI|PLSNET|PIDC|PUC|GRNVBEM|LEAP|NONLINEARODES|INFERELATOR]`: Inference techniques to be performed.  [required]\n* `--crossover-probability FLOAT`: Crossover probability  [default: 0.9]\n* `--num-parents INTEGER`: Number of parents  [default: 3]\n* `--mutation-probability FLOAT`: Mutation probability. [default: 1/len(files)]\n* `--mutation-strength FLOAT`: Mutation strength. [default: 0.1]\n* `--population-size INTEGER`: Population size  [default: 100]\n* `--num-evaluations INTEGER`: Number of evaluations  [default: 25000]\n* `--cut-off-criteria [MinConf|NumLinksWithBestConf|PercLinksWithBestConf]`: Criteria for determining which links will be part of the final binary matrix.  [default: PercLinksWithBestConf]\n* `--cut-off-value FLOAT`: Numeric value associated with the selected criterion. Ex: MinConf = 0.5, NumLinksWithBestConf = 10, PercLinksWithBestConf = 0.4  [default: 0.4]\n* `--function TEXT`: A mathematical expression that defines a particular fitness function based on the weighted sum of several independent terms. Available terms: Quality, DegreeDistribution and Motifs.  [required]\n* `--reference-point TEXT`: Reference point for the Pareto front. If specified, the search will be oriented towards this point. The format is \'f1;f2;f3\'. [default: "-"]\n* `--algorithm [GA|NSGAII|SMPSO]`: Evolutionary algorithm to be used during the optimization process. All are intended for a multi-objective approach with the exception of the genetic algorithm (GA).  [required]\n* `--threads INTEGER`: Number of threads to be used during parallelization. By default, the maximum number of threads available in the system is used.  [default: 64]\n* `--str-threads TEXT`: Comma-separated list with the identifying numbers of the threads to be used. If specified, the threads variable will automatically be set to the length of the list.\n* `--plot-fitness-evolution / --no-plot-fitness-evolution`: Indicate if you want to represent the evolution of the fitness values.  [default: no-plot-fitness-evolution]\n* `--plot-pareto-front / --no-plot-pareto-front`: Indicate if you want to represent the Pareto front (only available for multi-objective mode of 2 or 3 functions). [default: no-plot-pareto-front]\n* `--plot-parallel-coordinates / --no-plot-parallel-coordinates`: Indicate if you want to represent the parallel coordinate graph (only available for multi-objective mode). [default: no-plot-parallel-coordinates]\n* `--output-dir PATH`: Path to the output folder.  [default: inferred_networks]\n* `--help`: Show this message and exit.\n\n## `geneci optimize-ensemble`\n\nAnalyzes several trust lists and builds a consensus network by applying an evolutionary algorithm\n\n**Usage**:\n\n```console\n$ geneci optimize-ensemble [OPTIONS]\n```\n\n**Options**:\n\n* `--confidence-list TEXT`: Paths of the CSV files with the confidence lists to be agreed upon.  [required]\n* `--gene-names PATH`: Path to the TXT file with the name of the contemplated genes separated by comma and without space. If not specified, only the genes specified in the lists of trusts will be considered.\n* `--time-series PATH`: Path to the CSV file with the time series from which the individual gene networks have been inferred. This parameter is only necessary in case of specifying the fitness function Loyalty.\n* `--crossover-probability FLOAT`: Crossover probability  [default: 0.9]\n* `--num-parents INTEGER`: Number of parents  [default: 3]\n* `--mutation-probability FLOAT`: Mutation probability. [default: 1/len(files)]\n* `--mutation-strength FLOAT`: Mutation strength. [default: 0.1]\n* `--population-size INTEGER`: Population size  [default: 100]\n* `--num-evaluations INTEGER`: Number of evaluations  [default: 25000]\n* `--cut-off-criteria [MinConf|NumLinksWithBestConf|PercLinksWithBestConf]`: Criteria for determining which links will be part of the final binary matrix.  [default: PercLinksWithBestConf]\n* `--cut-off-value FLOAT`: Numeric value associated with the selected criterion. Ex: MinConf = 0.5, NumLinksWithBestConf = 10, PercLinksWithBestConf = 0.4  [default: 0.4]\n* `--function TEXT`: A mathematical expression that defines a particular fitness function based on the weighted sum of several independent terms. Available terms: Quality, DegreeDistribution and Motifs.  [required]\n* `--reference-point TEXT`: Reference point for the Pareto front. If specified, the search will be oriented towards this point. The format is \'f1;f2;f3\'. [default: "-"]\n* `--algorithm [GA|NSGAII|SMPSO]`: Evolutionary algorithm to be used during the optimization process. All are intended for a multi-objective approach with the exception of the genetic algorithm (GA).  [required]\n* `--threads INTEGER`: Number of threads to be used during parallelization. By default, the maximum number of threads available in the system is used.  [default: 64]\n* `--plot-fitness-evolution / --no-plot-fitness-evolution`: Indicate if you want to represent the evolution of the fitness values.  [default: no-plot-fitness-evolution]\n* `--plot-pareto-front / --no-plot-pareto-front`: Indicate if you want to represent the Pareto front (only available for multi-objective mode of 2 or 3 functions). [default: no-plot-pareto-front]\n* `--plot-parallel-coordinates / --no-plot-parallel-coordinates`: Indicate if you want to represent the parallel coordinate graph (only available for multi-objective mode). [default: no-plot-parallel-coordinates]\n* `--output-dir PATH`: Path to the output folder.  [default: <<conf_list_path>>/../ea_consensus]\n* `--help`: Show this message and exit.',
    'author': 'adrian_segura',
    'author_email': 'adrianseor.99@uma.es',
    'maintainer': 'None',
    'maintainer_email': 'None',
    'url': 'None',
    'packages': packages,
    'package_data': package_data,
    'install_requires': install_requires,
    'entry_points': entry_points,
    'python_requires': '>=3.9,<3.12',
}


setup(**setup_kwargs)
